lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -146,97 +145,117 @@ public void OnNext(IAllocatedEvaluator allocatedEvaluator)
         }
 
         /// <summary>
-        /// Specfies the Map or Update task to run on the active context
+        /// Specifies the Map or Update task to run on the active context
         /// </summary>
-        /// <param name="activeContext"></param>
-        public void OnNext(IActiveContext activeContext)
+        /// <param name="value"></param>
+        public void OnNext(IActiveContext value)
         {
-            Logger.Log(Level.Verbose, string.Format("Received Active Context {0}", activeContext.Id));
+            Logger.Log(Level.Verbose, string.Format("Received Active Context {0}", value.Id));
 
-            if (_groupCommDriver.IsMasterTaskContext(activeContext))
+            if (_groupCommDriver.IsMasterTaskContext(value))
             {
                 _reachedUpdateTaskActiveContext = true;
                 RequestMapEvaluators(_dataSet.Count);
+            }
 
-                var partialTaskConf =
-                    TangFactory.GetTang()
-                        .NewConfigurationBuilder(new[]
-                        {
-                            TaskConfiguration.ConfigurationModule
-                                .Set(TaskConfiguration.Identifier,
-                                    IMRUConstants.UpdateTaskName)
-                                .Set(TaskConfiguration.Task,
-                                    GenericType<UpdateTaskHost<TMapInput, TMapOutput, TResult>>.Class)
-                                .Build(),
-                            _configurationManager.UpdateFunctionConfiguration,
-                            _configurationManager.ResultHandlerConfiguration
-                        })
-                        .BindNamedParameter(typeof(InvokeGC), _invokeGC.ToString())
-                        .Build();
+            if (_contextManager.AddContext(value))
```

Comment:
I agree with @dkm2110, this would be much simpler if constructed as a series of observers, one of which acts as a guard for the gang scheduling constraint.